[5.5] [test] Use %refactor-check-compiles in more places

hamishknight opened the pull request:
5.5 cherry-pick of https://github.com/apple/swift/pull/37966, https://github.com/apple/swift/pull/38857 & https://github.com/apple/swift/pull/38915.

This is NFC, cherry-picking purely to avoid future merge conflicts in the event we pull more async refactoring changes into the 5.5 branch.

---

Add a hidden frontend flag `-warn-on-editor-placeholder` that allows `%refactor-check-compiles` to be more lenient with placeholders, and update a bunch of async refactoring tests to use `%refactor-check-compiles` in more cases.

Note the diff is best viewed without whitespace.

rdar://81830915

hamishknight commented:
@swift-ci please test

bnbarham commented:
@hamishknight `IF-LET-RETURN-CLASSIFICATION-NEXT` doesn't have a `IF-LET-RETURN-CLASSIFICATION` first in `path_classification.swift`. Seems that was fixed in `f5b00fc` on main.

hamishknight committed:
[test] Fix a couple of RUN directives

hamishknight committed:
[test] Re-enable convert_bool.swift

`%build-clang-importer-objc-overlays` builds the
overlays for the current target rather than the
host, so update the `%refactor` invocations to
take the target as an argument.

rdar://81128571

hamishknight committed:
[Frontend] Add -warn-on-editor-placeholder

This hidden frontend option lets us be more lax
when type-checking in the presence of editor
placeholders by treating them as holes during
constraint solving.

hamishknight committed:
Don't error on placeholders for refactor-check-compiles

This allows us to test more cases that type-check
modulo placeholders.

hamishknight committed:
Have refactor-check-compiles pass down -I -sdk -target

Forward these flags to both swift-frontend and
swift-refactor.

hamishknight committed:
[test] Use %refactor-check-compiles in more places

Update a bunch of async refactoring tests to use
%refactor-check-compiles in more cases.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows
